File: Week_02/G20200343030632/Leetcode_111_632.java
import java.util.Deque;
import java.util.LinkedList;


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        // bfs 广度优先
        Deque<TreeNode> queue = new LinkedList<>();
        int res = 1;
        if (root == null) {
            return 0;
        }
        queue.offer(root);
        while (!queue.isEmpty()) {
            int sz = queue.size();
            for (int i = 0; i < sz; i++) {
                TreeNode cur = queue.pollFirst();
                if (cur.left == null && cur.right == null) {
                    return res;
                }
                if (cur.left != null) {
                    queue.offerLast(cur.left);
                }
                if (cur.right != null) {
                    queue.offerLast(cur.right);
                }
            }
            res++;
        }
        return res;
    }
}

File: Week_02/G20200343030632/Leetcode_111_632.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minDepth(self, root: TreeNode) -> int:
        # return self.minDepth_v1(root)
        return self.minDepth_v2(root)
    
    def minDepth_v1(self, root: TreeNode) -> int:
        if not root:
            return 0
        if not root.left:
            return self.minDepth_v1(root.right) + 1
        if not root.right:
            return self.minDepth_v1(root.left) + 1
        return min(self.minDepth_v1(root.left), self.minDepth_v1(root.right)) + 1

    def minDepth_v2(self, root: TreeNode) -> int:
        if not root: return 0
        d = map(self.minDepth, (root.left, root.right))
        return 1 + (min(d) or max(d))
 